### PR TITLE
Restrict user access to specific applications

### PR DESCRIPTION
**Pull request by lgarvey**
## background
The Live Services team need the ability to control the level of access that user's have at the application level.  They want users to have no default access, which means that if a user is created in the `staff-sso` by default they will be returned to the application with an access denied error, until the LST grant them access. 

User's will be granted access to required applications when their accounts are bulk imported.

For any users who may have slipped through the net, there is a future task to create a page on the staff-sso broker informing the user that they do not yet have access and need to contact the Live Services Team.

## Implementation
After experimenting with scope backends and restricting scopes to applications and users, I decided to override the /o/authorize url in DOT and utilise the allow=True/False option of create_authorization_response(). By default if allow=False, the user is returned to the app's callback url with an ?error=access-denied query string appended.

This may have affected the alternative DOT oauth2 grant types. If this is the case, and we need to revive those types, a custom oauth2 application model with a field to revert to default DOT behaviour may be an option.

User's are granted access to applications using an M2M field defined on the User model.

### CODE REVIEW

**Comment by marcofucci** [comment]
Added @r4vi as well, I will review as soon as I can, sorry for the delay

File: sso/tests/test_user.py [hunk 1]
```diff
@@ -115,3 +117,25 @@ def test_get_short_name(self):
...
+        Test that `can_access()` returns True when user assigned to app
+        """
+
+        user = UserFactory()
+        app = ApplicationFactory(users=[user])
+
+        assert user.can_access(app)
+
+    @pytest.mark.django_db
+    def test_can_access_withut_perms(self):
```
Comment (marcofucci) [inline sso/tests/test_user.py]: did you mean _test_cannot_access_without_perms_?

File: sso/oauth2/views.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            # TODO: Cache this!
+            application = get_application_model().objects.get(client_id=credentials['client_id'])
+
+            kwargs['application'] = application
+            kwargs['client_id'] = credentials['client_id']
+            kwargs['redirect_uri'] = credentials['redirect_uri']
+            kwargs['response_type'] = credentials['response_type']
+            kwargs['state'] = credentials['state']
+
+            assert request.user.is_authenticated()
```
Comment (r4vi) [inline sso/oauth2/views.py]: is_authenticated as a function is depreciated, https://docs.djangoproject.com/en/1.11/ref/contrib/auth/#django.contrib.auth.models.User.is_authenticated
you can use it as a property `request.user.is_authenticated`

**Comment by lgarvey** [comment]
Well spotted! I'll correct the test name.

Agreed, it's a shame I had to copy verbatim - but I see no other way. =(

It may be we have to do some tests to confirm that both DH and DH-MI
integrations work with this PR before merging.

**Comment by r4vi** [comment]
@lgarvey if you need a list of users who should have access DH-MI we can provide it to you

**Pull request merged**